### PR TITLE
Feat: add deploy inline policy

### PR DESCRIPTION
**Pull request by Somefive**
Signed-off-by: Somefive <yd219913@alibaba-inc.com>

### Description of your changes

Add deploy inline policy and support loading definitions when lack in the revision.

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: charts/vela-core/templates/defwithtemplate/deploy.yaml [hunk 1]
```diff
@@ -21,15 +21,23 @@ spec:
         	policies:                 parameter.policies
         	parallelism:              parameter.parallelism
         	ignoreTerraformComponent: parameter.ignoreTerraformComponent
+        	inlinePolicies:           parameter.inlinePolicies
+        }
+        #Policy: {
+        	type:  string
+        	name?: string
+        	properties: {...}
```
Comment (FogDong) [inline charts/vela-core/templates/defwithtemplate/deploy.yaml:29]: Will the properties be empty?

Comment (Somefive) [reply charts/vela-core/templates/defwithtemplate/deploy.yaml:29]: > Will the properties be empty?

No, the policy type here needs to be one of "topology", "override", "replication". I have just added constraints on the policy type.

**Review by chivalryq — COMMENTED** [review]
If I get it right, we allow to `ParseWorkloadFromRevisionAndClient` because the inline override policy can introduce new Component/Trait? And the component/trait referred in non-inline policy have been added into AppRevision.

**Comment by Somefive** [comment]
> If I get it right, we allow to `ParseWorkloadFromRevisionAndClient` because the inline override policy can introduce new Component/Trait? And the component/trait referred in non-inline policy have been added into AppRevision.

Yes. Because inlinePolicies can be patched through workflow input/output so it is hard to detect definitions that needs to be used during the CreateAppRevision stage before workflow runs.

**Review by chivalryq — APPROVED** [review]

File: vela-templates/definitions/internal/workflowstep/deploy.cue [hunk 2]
```diff
@@ -15,15 +15,23 @@ template: {
...
 		auto: *true | bool
 		//+usage=Declare the policies that used for this deployment. If not specified, the components will be deployed to the hub cluster.
-		policies?: [...string]
+		policies: *[] | [...string]
 		//+usage=Maximum number of concurrent delivered components.
 		parallelism: *5 | int
 		//+usage=If set false, this step will apply the components with the terraform workload.
 		ignoreTerraformComponent: *true | bool
+		//+usage=The policies that embeds in the `deploy` step directly
+		inlinePolicies: *[] | [...#Policy]
```
Comment (barnettZQG) [inline vela-templates/definitions/internal/workflowstep/deploy.cue:35]: Adding this field for the `deploy` step may not be a good idea. This leads the user to choose between the `policies` and `inlinePolicies` fields. The `policies` field is recommended. 

But, If we want to define an `xx-deploy` step type with some default policies, this field is very useful.

Comment (Somefive) [reply vela-templates/definitions/internal/workflowstep/deploy.cue:35]: I see. So you recommend not exposing inlinePolicies in deploy.cue but in op.#Deploy right?

Comment (barnettZQG) [reply vela-templates/definitions/internal/workflowstep/deploy.cue:35]: Yes

Comment (Somefive) [reply vela-templates/definitions/internal/workflowstep/deploy.cue]: Updated.

**Review by barnettZQG — APPROVED** [review]
LGTM

**Review by FogDong — APPROVED** [review]

**Pull request merged**